systemtest/infra.go

```diff
@@ -0,0 +1,119 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package systemtest
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"os"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/hc-install/product"
+	"github.com/hashicorp/hc-install/releases"
+	"github.com/hashicorp/terraform-exec/tfexec"
+)
+
+// NewTerraform returns a terraform executable for the specified path. It
+// installs the latest version.
+func NewTerraform(ctx context.Context, path string) (*tfexec.Terraform, error) {
+	execPath, err := installTerraform(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return tfexec.NewTerraform(path, execPath)
+}
+
+var terraformExecMu sync.Mutex
+var terraformExec string
+
+func installTerraform(ctx context.Context) (string, error) {
```

Comment:
**suggestion**: up to you but what about using a `sync.Once` to remove the lock and the two if conditions ?